### PR TITLE
OSPFv2_interfaces - Add Enterprise SONiC resource module

### PR DESCRIPTION
**Pull request by santhosh-kt**
##### SUMMARY

- Add Enterprise SONiC 'OSPFv2_interfaces' resource module:
    - Provide data definitions in 'argspec'.
    - Implement 'facts' to gather facts for the module.
    - Implement 'config' to apply configurations for `merged` , `deleted`, `overridden` and `replaced` states.
- Define regression test cases to verify correct functionality of ospfv2_interfaces resource module.
- Define unit test cases to verify correct functionality of ospfv2_interfaces resource module.

##### GitHub Issues
List the GitHub issues impacted by this PR. If no Github issues are affected, please indicate this with "N/A".

| GitHub Issue # |
| -------------- |
| |

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- sonic_ospfv2_interfaces

##### OUTPUT

[regression-2024-08-05-14-48-41.zip](https://github.com/user-attachments/files/16493865/regression-2024-08-05-14-48-41.zip)

##### ADDITIONAL INFORMATION

```paste below

```

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: plugins/modules/sonic_ospfv2_interfaces.py [hunk 1]
```diff
@@ -0,0 +1,748 @@
...
+                required: True
+              pwd:
+                description:
+                  - Configure OSPFv2 message digest password.
+                  - Plain text passwords i.e. password with encrypted=False will be encrypted in running-config, so idempotency will
+                    not be maintained and every time the output will come as changed=True.
+                type: str
+              encrypted:
+                description:
+                  - Indicates whether the password is encrypted text.
```
Comment (ArunSaravananBalachandran) [inline plugins/modules/sonic_ospfv2_interfaces.py:139]:
```suggestion
          message_digest_key:
            description:
              - Configure OSPFv2 message digest keys and password.
              - Uses MD5 algorithm.
            type: list
            elements: dict
            suboptions:
              key_id:
                description:
                  - Specifies the OSPFv2 message digest key ID (1 to 255).
                type: int
                required: True
              md5key:
                description:
                  - Specifies the OSPFv2 message digest password.
                  - Plain text password i.e. password with I(encrypted=false) will be stored in encrypted format in running-config, so idempotency will
                   not be maintained and hence the task output will always be I(changed=true).
                type: str
              encrypted:
                description:
                  - Indicates whether the password is in encrypted format.
```

File: plugins/module_utils/network/sonic/facts/ospfv2_interfaces/ospfv2_interfaces.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                                        self.update_dict(attr['auth_pwd'], 'encrypted', config.get('authentication-key-encrypted'))
+
+                                    if 'dead-interval-minimal' in config:
+                                        dead_interval_minimal = config.get('dead-interval-minimal')
+                                        if dead_interval_minimal:
+                                            self.update_dict(attr, 'hello_multiplier', config.get('hello-multiplier'))
+                                        else:
+                                            self.update_dict(attr, 'dead_interval', config.get('dead-interval'))
+                                    elif 'dead-interval' in config:
+                                        self.update_dict(attr, 'dead_interval', config.get('dead-interval'))
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/facts/ospfv2_interfaces/ospfv2_interfaces.py:160]: Could be simplified as:
```suggestion
                                    if config.get('dead-interval-minimal'):
                                        self.update_dict(attr, 'hello_multiplier', config.get('hello-multiplier'))
                                    else:
                                        self.update_dict(attr, 'dead_interval', config.get('dead-interval'))
```

File: plugins/module_utils/network/sonic/facts/ospfv2_interfaces/ospfv2_interfaces.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                                            md_key = {}
+                                            self.update_dict(md_key, 'key_id', md_config.get('authentication-key-id'))
+                                            self.update_dict(md_key, 'pwd', md_config.get('authentication-md5-key'))
+                                            self.update_dict(md_key, 'encrypted', md_config.get('authentication-key-encrypted'))
+                                            if md_key:
+                                                md_keys.append(md_key)
+
+                                    self.update_dict(attr, 'message_digest_pwd', md_keys)
+                                if attr:
+                                    if addr != '0.0.0.0':
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/facts/ospfv2_interfaces/ospfv2_interfaces.py:180]: Please use a variable for the default interface OSPF address.

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

**Comment by jeff-yin** [comment]
@ArunSaravananBalachandran  please review the latest changes, resolve any open comments that you feel have been adequately addressed, and review/approve this PR.

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

File: plugins/modules/sonic_ospfv2_interfaces.py [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+"""
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+---
+module: sonic_ospfv2_interfaces
+version_added: '3.0.0'
```
Comment (ArunSaravananBalachandran) [inline plugins/modules/sonic_ospfv2_interfaces.py:36]:
```suggestion
version_added: '2.5.0'
```

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

File: plugins/module_utils/network/sonic/config/ospfv2_interfaces/ospfv2_interfaces.py [hunk 5]
```diff
@@ -0,0 +1,761 @@
...
+                add_cfg, del_cfg = {}, {}
+                for attr in OSPF_INT_ATTRIBUTES:
+                    if attr in conf:
+                        if attr not in have_conf:
+                            add_cfg[attr] = conf[attr]
+                        else:
+                            if attr == 'bfd':
+                                for bfd_attr in ['enable', 'bfd_profile']:
+                                    if bfd_attr in conf[attr]:
+                                        if bfd_attr not in have_conf:
```
Comment (ArunSaravananBalachandran) [inline plugins/module_utils/network/sonic/config/ospfv2_interfaces/ospfv2_interfaces.py:297]:
```suggestion
                                        if bfd_attr not in have_conf[attr]:
```

**Review by ArunSaravananBalachandran — CHANGES_REQUESTED** [review]

**Review by ArunSaravananBalachandran — APPROVED** [review]

**Review by kerry-meyer — APPROVED** [review]
The posted change set and corresponding test results look good.

Thank you very much for implementing the OSPFv2 "interfaces" resource module for the enterprise_sonic Ansible collection.

Approved.

**Pull request merged**